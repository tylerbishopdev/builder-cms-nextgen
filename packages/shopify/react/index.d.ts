export * from '../dist/react/react';
