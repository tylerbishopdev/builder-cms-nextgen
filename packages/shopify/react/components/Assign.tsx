import { BuilderStore, onChange, withBuilder } from '@builder.io/react';
import * as React from 'react';
import { AssignBlockProps } from '../interfaces/component-props';

export class AssignBlock extends React.Component<AssignBlockProps> {
  ran = false;

  constructor(props: AssignBlockProps) {
    super(props);
  }

  run() {
    const { expression, builderState } = this.props;

    if (expression && builderState) {
      if (builderState.context.shopify) {
        builderState.context.shopify.liquid.assign(expression, onChange.target(builderState.state));
        this.ran = true;
        return true;
      }
    }
    return false;
  }

  render() {
    this.run();
    return null;
  }
}

withBuilder(AssignBlock, {
  name: 'Shopify:Assign',
  hideFromInsertMenu: true,
  noWrap: true,
  inputs: [
    {
      name: 'expression',
      type: 'javascript',
      // TODO: type: 'code', language: 'javascript'
    },
  ],
});
