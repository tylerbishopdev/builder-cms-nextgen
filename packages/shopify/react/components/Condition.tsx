import builder, {
  Builder,
  BuilderBlockComponent,
  BuilderElement,
  BuilderStore,
  stringToFunction,
  withBuilder,
} from '@builder.io/react';
import * as React from 'react';
import { ConditionBlockProps } from '../interfaces/component-props';

export class ConditionBlock extends React.Component<ConditionBlockProps> {
  private getMatchingBranch(branches = this.props.branches) {
    if (!(branches && branches.length)) {
      return null;
    }

    for (let i = 0; i < branches.length; i++) {
      const branch = branches[i];
      const isLast = i === branches.length - 1;
      if (isLast && !branch.expression) {
        return { index: i, blocks: branch.blocks };
      }
      const fn = stringToFunction(branch.expression || '');
      const result = fn(
        this.props.builderState && this.props.builderState.state,
        null,
        this.props.builderBlock,
        builder,
        null,
        this.props.builderState!.update,
        Builder,
        this.props.builderState?.context
      );
      if (result) {
        return { index: i, blocks: branch.blocks };
      }
    }
    return null;
  }

  render() {
    const result = this.getMatchingBranch();
    if (!result) {
      return null;
    }

    return result.blocks.map(block => <BuilderBlockComponent key={block.id} block={block} />);
  }
}

const defaultBlock = {
  '@type': '@builder.io/sdk:Element',
  responsiveStyles: {
    large: {
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'stretch',
      flexShrink: '0',
      position: 'relative',
      marginTop: '30px',
      textAlign: 'center',
      lineHeight: 'normal',
      height: 'auto',
    },
  },
  component: {
    name: 'Text',
    options: {
      text: '<p>Enter some text...</p>',
    },
  },
};

withBuilder(ConditionBlock, {
  name: 'Shopify:Condition',
  hideFromInsertMenu: true,
  noWrap: true,
  inputs: [
    {
      name: 'branches',
      type: 'array',
      subFields: [
        {
          name: 'expression',
          type: 'javascript',
        },
        {
          name: 'blocks',
          type: 'array',
          hideFromUI: true,
          defaultValue: [defaultBlock],
        },
      ],
    },
  ],
});
