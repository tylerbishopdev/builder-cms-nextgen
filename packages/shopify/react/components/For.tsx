import {
  BuilderStore,
  withBuilder,
  BuilderStoreContext,
  BuilderBlockComponent,
  BuilderElement,
  Builder,
} from '@builder.io/react';
import React, { useState } from 'react';

interface ForBlockProps {
  repeat: {
    itemName: string;
    collection: string;
  };
  builderState?: BuilderStore;
  builderBlock?: BuilderElement;
}

export function ForBlock(props: ForBlockProps) {
  let arrayItems: any[] = [];
  const { repeat, builderState } = props;

  const collectionPath = repeat.collection;
  const collectionName = (collectionPath || '').trim().split('(')[0].trim().split('.').slice(-1)[0];

  const itemName = repeat.itemName || (collectionName ? collectionName + 'Item' : 'item');

  if (itemName && collectionName && builderState && Builder.isBrowser) {
    if (builderState.context.shopify) {
      const renderedItems = builderState.context.shopify.liquid.get(
        collectionPath,
        builderState.state
      );

      if (Array.isArray(renderedItems)) {
        arrayItems = renderedItems;
      }
    }
  }

  return arrayItems?.map((item, index) => {
    const scopedState: any = {
      forloop: {
        index0: index,
        index: index + 1,
        first: index === 0,
        last: index === arrayItems.length - 1,
        length: arrayItems.length,
        rindex: arrayItems.length - index,
        rindex0: arrayItems.length - index - 1,
      },
    };

    // add the itemName (ie "product") to the state so the children can access
    scopedState[itemName] = item;

    return (
      <BuilderStoreContext.Consumer key={index}>
        {store => (
          <BuilderStoreContext.Provider
            value={{
              ...store,
              state: {
                ...store.state,
                ...scopedState,
              },
            }}
          >
            {props.builderBlock?.children?.map(item => (
              <BuilderBlockComponent block={item} key={item.id} />
            ))}
          </BuilderStoreContext.Provider>
        )}
      </BuilderStoreContext.Consumer>
    );
  });
}

withBuilder(ForBlock, {
  name: 'Shopify:For',
  hideFromInsertMenu: true,
  noWrap: true,
  inputs: [
    {
      name: 'expression',
      type: 'javascript',
    },
  ],
});
