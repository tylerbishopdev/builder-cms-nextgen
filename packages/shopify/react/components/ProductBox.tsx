/**
 * IMPORTANT: THIS FILE IS DEPRECATED
 * use product-box.ts instead
 */

/** @jsx jsx */
import { jsx, css, keyframes } from '@emotion/core';
import styled from '@emotion/styled';
import { ShopifyProduct } from '../interfaces/shopify-product';
import React, { useEffect, useState } from 'react';
import { BuilderElement } from '@builder.io/sdk';
import {
  Builder,
  BuilderStore,
  BuilderStoreContext,
  BuilderBlockComponent,
} from '@builder.io/react';
import { modifyProduct } from '../functions/modify-product';

interface ProductBoxProps {
  product?: string | number;
  linkToProductPageOnClick?: boolean;
  /** @todo implement this */
  variant?: string | number;
  builderState?: BuilderStore;
  builderBlock?: BuilderElement;
}

const spin = keyframes`
  0% {
    transform:rotate(0)
  }
  100% {
    transform:rotate(360deg)
  }
`;

const Box = styled.div({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'stretch',
});

const LoadingSpinner = styled.div`
  width: 4em;
  height: 4em;
  border-radius: 50%;
  position: relative;
  margin: 6rem auto;
  font-size: 1rem;
  text-indent: -9999em;
  border-top: 0.2em solid rgba(131, 132, 137, 0.2);
  border-right: 0.2em solid rgba(131, 132, 137, 0.2);
  border-bottom: 0.2em solid rgba(131, 132, 137, 0.2);
  border-left: 0.2em solid #454749;
  transform: translateZ(0);
  animation: 1.1s linear infinite ${spin};
`;

/**
 * @deprecated, use product-box.ts symbol instead
 */
export function ProductBox(props: ProductBoxProps) {
  let productId = props.product || '';
  const [productInfo, setProductInfo] = useState<ShopifyProduct | null>(null);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    updateProduct();
  }, [props.product]);

  function updateProduct() {
    const { context } = props.builderState!;

    if (productId) {
      setLoading(true);
      fetch(
        `https://cdn.builder.io/api/v1/shopify/products/${productId}.json?apiKey=${context.apiKey}`
      )
        .then(res => res.json())
        .then(data => {
          const { product } = data;
          if (product) {
            const modifiedProduct = modifyProduct(product);
            setProductInfo(modifiedProduct);
          }
          setLoading(false);
        })
        .catch(err => {
          console.error('Error fetching Shopify product', err);
          setLoading(false);
        });
    } else {
      setProductInfo(null);
    }
  }

  function onClick(e: React.MouseEvent<HTMLDivElement, MouseEvent>) {
    if (!props.builderBlock?.component?.options?.linkToProductPageOnClick) {
      return;
    }
    if (!productInfo) {
      return;
    }
    if (e.defaultPrevented || Builder.isEditing) {
      return;
    }
    const url = `/products/${productInfo.handle}`;
    function goToProductPage() {
      location.href = url;
    }
    const { target, currentTarget } = e;
    if (target === currentTarget) {
      goToProductPage();
    }
    let current = target as HTMLElement | null;
    do {
      if (!current) {
        break;
      } else if (current === currentTarget) {
        goToProductPage();
      } else if (['SELECT', 'INPUT', 'FORM', 'BUTTON', 'A'].includes(current.tagName)) {
        break;
      }
    } while (current && (current = current.parentElement));
  }

  return Builder.isEditing && !productId ? (
    // TODO: <Info> component
    <div
      style={{
        padding: 20,
        opacity: 0.7,
        textAlign: 'center',
      }}
    >
      Double click to choose a product
    </div>
  ) : loading && !productInfo ? (
    <LoadingSpinner />
  ) : (
    <BuilderStoreContext.Consumer>
      {store => (
        <BuilderStoreContext.Provider
          value={{
            ...store,
            state: {
              ...store.state,
              // TODO: communicate up this up to DataEditor to populate bindings dropdowns (how?)
              productInfo,
            },
          }}
        >
          <Box
            css={props.linkToProductPageOnClick ? { cursor: 'pointer' } : undefined}
            onClick={onClick}
          >
            {props.builderBlock?.children?.map(item => (
              <BuilderBlockComponent block={item} key={item.id} />
            ))}
          </Box>
        </BuilderStoreContext.Provider>
      )}
    </BuilderStoreContext.Consumer>
  );
}

Builder.registerComponent(ProductBox, {
  name: 'Shopify:ProductBox',
  image:
    'https://cdn.builder.io/api/v1/image/assets%2Fd4fb2d0e6f954b31b60f6f0573e91b0c%2Fff2fd04bad384f30837b40bda8a062dc',
  canHaveChildren: true,
  hideFromInsertMenu: true,
  inputs: [
    {
      name: 'product',
      type: 'ShopifyProduct',
    },
    {
      name: 'linkToProductPageOnClick',
      type: 'boolean',
      defaultValue: true,
    },
  ],
});
