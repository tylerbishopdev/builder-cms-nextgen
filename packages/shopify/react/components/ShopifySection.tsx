import React from 'react';
import { BuilderElement } from '@builder.io/sdk';
import { StateProvider, withBuilder } from '@builder.io/react';

interface StateProviderProps {
  builderBlock: BuilderElement;
  state: any;
  schema: any;
}

const ShopifySectionComponent: React.FC<StateProviderProps> = props => {
  const { schema, ...stateProviderProps } = props;
  return (
    <div
      id={`shopify-section-${schema.class?.replace('-section', '')}`}
      className={`shopify-section ${schema.class}`}
    >
      <StateProvider {...stateProviderProps} />
    </div>
  );
};

export const ShopifySection = withBuilder(ShopifySectionComponent, {
  name: 'Shopify:Section',
  canHaveChildren: true,
  static: true,
  hideFromInsertMenu: true,
});
