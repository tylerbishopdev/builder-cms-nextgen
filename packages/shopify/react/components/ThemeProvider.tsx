import React from 'react';
import { BuilderElement } from '@builder.io/sdk';
import { StateProvider, withBuilder } from '@builder.io/react';

interface ThemeProviderProps {
  builderBlock: BuilderElement;
  state: any;
}

const ThemeProviderComponent: React.FC<ThemeProviderProps> = props => <StateProvider {...props} />;

export const ThemeProvider = withBuilder(ThemeProviderComponent, {
  name: 'Shopify:ThemeProvider',
  canHaveChildren: true,
  static: true,
  hideFromInsertMenu: true,
});
