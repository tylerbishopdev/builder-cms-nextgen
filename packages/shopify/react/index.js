export * from '../dist/react/bundle.esm.js';
