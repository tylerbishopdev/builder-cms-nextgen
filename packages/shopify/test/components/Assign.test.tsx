import { render } from '@testing-library/react';
import * as React from 'react';
import { AssignBlock } from '../../react/components/Assign';
import { mockStateWithShopify } from '../modules/helpers';
import * as reactTestRenderer from 'react-test-renderer';

describe('Assign', () => {
  test('Basic assignment', () => {
    const state: any = {};
    const ref = render(
      <AssignBlock expression="foo = 'bar'" builderState={mockStateWithShopify(state)} />
    );
    expect(state.foo).toEqual('bar');
  });

  test('Multi assignment', () => {
    const state: any = {};
    const ref = render(
      <>
        <AssignBlock expression="foo = 'bar'" builderState={mockStateWithShopify(state)} />
        <AssignBlock expression="bar = foo" builderState={mockStateWithShopify(state)} />
      </>
    );
    expect(state.bar).toEqual('bar');
  });

  it('renders snapshot correctly', () => {
    const state: any = {};
    const tree = reactTestRenderer
      .create(<AssignBlock expression="foo = 'bar'" builderState={mockStateWithShopify(state)} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
