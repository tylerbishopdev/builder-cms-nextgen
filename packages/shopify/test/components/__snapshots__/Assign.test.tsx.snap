// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assign renders snapshot correctly 1`] = `null`;
